src/cascadia/WindowsTerminal/ContentProcessMain.cpp

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "AppHost.h"
+#include "resource.h"
+#include "../types/inc/User32Utils.hpp"
+#include <WilErrorReporting.h>
+
+using namespace winrt;
+using namespace winrt::Windows::UI;
+using namespace winrt::Windows::UI::Composition;
+using namespace winrt::Windows::UI::Xaml::Hosting;
+using namespace winrt::Windows::Foundation::Numerics;
+
+// We keep a weak ref to our ContentProcess singleton here.
+// Why?
+//
+// We need to always return the _same_ ContentProcess when someone comes to
+// instantiate this class. So we want to track the single instance we make. We
+// also want to track when the last outstanding reference to this object is
+// removed. If we're keeping a strong ref, then the ref count will always be > 1
+
+winrt::weak_ref<winrt::Microsoft::Terminal::Control::ContentProcess> g_weak{ nullptr };
+wil::unique_event g_canExitThread;
+
+struct ContentProcessFactory : implements<ContentProcessFactory, IClassFactory>
+{
+    ContentProcessFactory(winrt::guid g) :
+        _guid{ g } {};
+
+    HRESULT __stdcall CreateInstance(IUnknown* outer, GUID const& iid, void** result) noexcept final
+    {
+        *result = nullptr;
+        if (outer)
+        {
+            return CLASS_E_NOAGGREGATION;
+        }
+
+        if (!g_weak)
+        {
+            // Instantiate the ContentProcess here
+            winrt::Microsoft::Terminal::Control::ContentProcess strong{};
+
+            // Now, create a weak ref to that ContentProcess object.
+            winrt::weak_ref<winrt::Microsoft::Terminal::Control::ContentProcess> weak{ strong };
+
+            // Stash away that weak ref for future callers.
+            g_weak = weak;
+            return strong.as(iid, result);
+        }
+        else
+        {
+            auto strong = g_weak.get();
+            // !! LOAD BEARING !! If you set this event in the _first_ branch
+            // here, when we first create the object, then there will be _no_
+            // references to the ContentProcess object for a small slice. We'll
+            // stash the ContentProcess in the weak_ptr, and return it, and at
+            // that moment, there will be 0 outstanding references, it'll dtor,
+            // and we'll ExitProcess.
+            //
+            // Instead, set the event here, once there's already a reference
+            // outside of just the weak one we keep. Experimentation showed this
+            // waw always hit when creating the ContentProcess at least once.

```

Comment:
I wonder why `waw` didn't trip our spell checker
```suggestion
            // was always hit when creating the ContentProcess at least once.
```